### PR TITLE
Pass regexp filter to /stats/prometheus

### PR DESCRIPTION
**Pull request by StrongMonkey**
Description: Pass regexp parameter to /stats/prometheus handler
Risk Level: low
Issue: https://github.com/envoyproxy/envoy/issues/7173

### CODE REVIEW

**Comment by StrongMonkey** [comment]
@mattklein123 Hi, Can you help review this pr?

File: source/server/http/admin.cc [hunk 1]
```diff
@@ -753,8 +753,13 @@ Http::Code AdminImpl::handlerPrometheusStats(absl::string_view path_and_query, H
                                              Buffer::Instance& response, AdminStream&) {
   const Http::Utility::QueryParams params = Http::Utility::parseQueryString(path_and_query);
   const bool used_only = params.find("usedonly") != params.end();
+  const absl::optional<std::regex> regex =
+      (params.find("filter") != params.end())
+          ? absl::optional<std::regex>{std::regex(params.at("filter"))}
```
Comment (snowp) [inline source/server/http/admin.cc]: I'd probably split this up into multiple lines for the sake of readability

File: source/server/http/admin.h [hunk 2]
```diff
@@ -437,6 +438,12 @@ class PrometheusStatsFormatter {
   static bool shouldShowMetric(const std::shared_ptr<Stats::Metric>& metric, const bool used_only) {
```
Comment (snowp) [inline source/server/http/admin.h]: Is this still used?

**Comment by StrongMonkey** [comment]
@snowp Hi, thanks for reviewing! I pushed my code to address your comments, and also added a unit test to exercise the regex for prometheus stats.

File: source/server/http/admin.cc (hunk 1 above)
Comment (mattklein123) [reply source/server/http/admin.cc]: Drive by, this is already computed above in `handlerStats()`. I would just pass that computation to this function.

Comment (StrongMonkey) [reply source/server/http/admin.cc]: @mattklein123 I was actually looking at https://github.com/envoyproxy/envoy/blob/89a82de7a8368cd93b1255dfb3818f497330faf2/source/server/http/admin.cc#L1208 and seems like there is no good way unless we want to change the definition of MAKE_ADMIN_HANDLER. It is when the method is called from top-level api handler.

Comment (mattklein123) [reply source/server/http/admin.cc]: I see, OK. Maybe just make a smaller helper method?

Comment (StrongMonkey) [reply source/server/http/admin.cc]: Helper method added.

File: source/server/http/admin.h [hunk 3]
```diff
@@ -220,6 +220,11 @@ class AdminImpl : public Admin,
                                  bool used_only,
                                  const absl::optional<std::regex> regex = absl::nullopt,
                                  bool pretty_print = false);
+  static absl::optional<std::regex> filterParam(Http::Utility::QueryParams params) {
```
Comment (snowp) [inline source/server/http/admin.h]: Minor thing but I'd probably put this in the anonymous namespace in the .cc file, no reason to put this on the class

Comment (StrongMonkey) [reply source/server/http/admin.h]: updated

File: source/server/http/admin.h [hunk 4]
```diff
@@ -413,7 +414,8 @@ class PrometheusStatsFormatter {
   static uint64_t statsAsPrometheus(const std::vector<Stats::CounterSharedPtr>& counters,
                                     const std::vector<Stats::GaugeSharedPtr>& gauges,
                                     const std::vector<Stats::ParentHistogramSharedPtr>& histograms,
-                                    Buffer::Instance& response, const bool used_only);
+                                    Buffer::Instance& response, const bool used_only,
+                                    const absl::optional<std::regex> regex);
```
Comment (mattklein123) [inline source/server/http/admin.h]: nit: `const absl::optional<std::regex>&`

Comment (StrongMonkey) [reply source/server/http/admin.h]: thanks, updated!

**Pull request merged**